<a href="https://colab.research.google.com/github/jamesqiu2021/DSCI-401-QW/blob/main/HW3_QW_submit.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import pandas as pd
airlines = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/airlines.csv")
flights = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/flights.csv")
airports = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/airports.csv")
batting = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/Batting.csv")
people = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/People.csv")
Pitching = pd.read_csv("https://raw.githubusercontent.com/jamesqiu2021/DSCI-401-QW/main/Pitching.csv")
planes = pd.read_csv("https://raw.githubusercontent.com/tidyverse/nycflights13/main/data-raw/planes.csv")
pccc = pd.read_csv("https://raw.githubusercontent.com/gjm112/DSCI401/main/data/pccc_icd10_dataset.csv")


#1(a) (10 points) Name every player in baseball history who has accumulated at least 300 home runs (HR) and at least 300 stolen bases (SB).You can find the first and last name of the player in the People data frame. Join this to your result along with the total home runs and total bases stolen for each of these elite players.

In [3]:
from numpy.lib.shape_base import column_stack
column_stack(people)

array([['playerID', 'birthYear', 'birthMonth', 'birthDay',
        'birthCountry', 'birthState', 'birthCity', 'deathYear',
        'deathMonth', 'deathDay', 'deathCountry', 'deathState',
        'deathCity', 'nameFirst', 'nameLast', 'nameGiven', 'weight',
        'height', 'bats', 'throws', 'debut', 'finalGame', 'retroID',
        'bbrefID', 'deathDate', 'birthDate']], dtype='<U12')

In [4]:
from pandas.core.resample import PeriodIndexResampler
from pandas.core.reshape.merge import merge
q = batting.loc[:,["playerID","HR","SB"]].groupby("playerID").sum()
q1=q[q.HR>300][q.SB>300]
people["name"]=people["nameFirst"]+" " + people["nameLast"]
people_1=people[["playerID","name"]]
merge_q = pd.merge(people_1,q1,on="playerID",how="inner")
merge_q




/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:4: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  after removing the cwd from sys.path.


playerID            name   HR     SB
0  beltrca01  Carlos Beltran  435  312.0
1  bondsba01     Barry Bonds  762  514.0
2  bondsbo01     Bobby Bonds  332  461.0
3  dawsoan01    Andre Dawson  438  314.0
4  finlest01    Steve Finley  304  320.0
5   mayswi01     Willie Mays  660  338.0
6  rodrial01  Alex Rodriguez  696  329.0
7  sandere02  Reggie Sanders  305  304.0

#1(b) (10 points) Similarly, name every pitcher in baseball history who has accumulated at least 300 wins (W) and at least 3,000 strikeouts (SO).

In [5]:
q2 = Pitching.loc[:,["playerID","W","SO"]].groupby("playerID").sum()
q3=q2[q2.W>300][q2.SO>3000]
people["name"]=people["nameFirst"]+" "+ people["nameLast"]
people_1=people[["playerID","name"]]
merge_q2 = pd.merge(people_1,q3,on="playerID",how="inner")
merge_q2

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  


playerID            name    W    SO
0  carltst01   Steve Carlton  329  4136
1  clemero02   Roger Clemens  354  4672
2  johnsra05   Randy Johnson  303  4875
3  johnswa01  Walter Johnson  417  3509
4  maddugr01     Greg Maddux  355  3371
5  niekrph01     Phil Niekro  318  3342
6  perryga01   Gaylord Perry  314  3534
7   ryanno01      Nolan Ryan  324  5714
8  seaveto01      Tom Seaver  311  3640
9  suttodo01      Don Sutton  324  3574

#1(c) (10 points) Identify the name and year of every player who has hit at least 50 home runs in a single season. Which player had the lowest batting average in that season?

In [6]:
q1c1_c=batting[["yearID","playerID","HR"]].groupby(["yearID","playerID"],as_index = False).sum()
q1c1_c=q1c1_c[q1c1_c.HR>=50].sort_values("HR",ascending=False)
q1c1_c

yearID   playerID  HR
74980    2001  bondsba01  73
71933    1998  mcgwima01  70
72243    1998   sosasa01  66
73138    1999  mcgwima01  65
75909    2001   sosasa01  64
73459    1999   sosasa01  63
38652    1961  marisro01  61
20647    1927   ruthba01  60
96427    2017  stantmi03  59
17567    1921   ruthba01  59
22913    1932   foxxji01  58
25842    1938  greenha01  58
81538    2006  howarry01  58
70784    1997  mcgwima01  58
77031    2002  rodrial01  57
75275    2001  gonzalu01  57
71641    1998  griffke02  56
22270    1930  wilsoha01  56
70513    1997  griffke02  56
81848    2006  ortizda01  54
83255    2007  rodrial01  54
86163    2010  bautijo02  54
17076    1920   ruthba01  54
21155    1928   ruthba01  54
31850    1949  kinerra01  54
38650    1961  mantlmi01  54
98026    2019  alonspe01  53
90210    2013  davisch02  53
75820    2001  rodrial01  52
95885    2017  judgeaa01  52
77177    2002  thomeji01  52
35752    1956  mantlmi01  52
69643    1996  mcgwima01  52
41403    1965   mayswi01  52
50772    1977  fostege01  52
80368    2005  jonesan01  51
30864    1947   mizejo01  51
30769    1947  kinerra01  51
62963    1990  fieldce01  51
35184    1955   mayswi01  51
69007    1996  anderbr01  50
74668    2000   sosasa01  50
82625    2007  fieldpr01  50
25815    1938   foxxji01  50
72334    1998  vaughgr01  50
67920    1995  belleal01  50

In [7]:
merge_q3 = pd.merge(q1c1_c,people_1,on="playerID",how="inner")
merge_q3

yearID   playerID  HR               name
0     2001  bondsba01  73        Barry Bonds
1     1998  mcgwima01  70       Mark McGwire
2     1999  mcgwima01  65       Mark McGwire
3     1997  mcgwima01  58       Mark McGwire
4     1996  mcgwima01  52       Mark McGwire
5     1998   sosasa01  66         Sammy Sosa
6     2001   sosasa01  64         Sammy Sosa
7     1999   sosasa01  63         Sammy Sosa
8     2000   sosasa01  50         Sammy Sosa
9     1961  marisro01  61        Roger Maris
10    1927   ruthba01  60          Babe Ruth
11    1921   ruthba01  59          Babe Ruth
12    1920   ruthba01  54          Babe Ruth
13    1928   ruthba01  54          Babe Ruth
14    2017  stantmi03  59  Giancarlo Stanton
15    1932   foxxji01  58        Jimmie Foxx
16    1938   foxxji01  50        Jimmie Foxx
17    1938  greenha01  58     Hank Greenberg
18    2006  howarry01  58        Ryan Howard
19    2002  rodrial01  57     Alex Rodriguez
20    2007  rodrial01  54     Alex Rodriguez
21    2001  rodrial01  52     Alex Rodriguez
22    2001  gonzalu01  57      Luis Gonzalez
23    1998  griffke02  56        Ken Griffey
24    1997  griffke02  56        Ken Griffey
25    1930  wilsoha01  56        Hack Wilson
26    2006  ortizda01  54        David Ortiz
27    2010  bautijo02  54      Jose Bautista
28    1949  kinerra01  54        Ralph Kiner
29    1947  kinerra01  51        Ralph Kiner
30    1961  mantlmi01  54      Mickey Mantle
31    1956  mantlmi01  52      Mickey Mantle
32    2019  alonspe01  53        Pete Alonso
33    2013  davisch02  53        Chris Davis
34    2017  judgeaa01  52        Aaron Judge
35    2002  thomeji01  52          Jim Thome
36    1965   mayswi01  52        Willie Mays
37    1955   mayswi01  51        Willie Mays
38    1977  fostege01  52      George Foster
39    2005  jonesan01  51       Andruw Jones
40    1947   mizejo01  51        Johnny Mize
41    1990  fieldce01  51      Cecil Fielder
42    1996  anderbr01  50     Brady Anderson
43    2007  fieldpr01  50     Prince Fielder
44    1998  vaughgr01  50        Greg Vaughn
45    1995  belleal01  50       Albert Belle

In [8]:
merge_q4 = pd.merge(merge_q3,batting,on=("playerID","yearID"),how="inner")
merge_q4["batting_average"]=merge_q4["H"]/merge_q4["AB"]
merge_q4[["yearID","playerID","name","batting_average"]].sort_values(by="batting_average",ascending=True).head(5)

yearID   playerID           name  batting_average
4     1997  mcgwima01   Mark McGwire         0.252874
33    2019  alonspe01    Pete Alonso         0.259631
28    2010  bautijo02  Jose Bautista         0.260105
40    2005  jonesan01   Andruw Jones         0.262799
10    1961  marisro01    Roger Maris         0.269492

#2. Use the nycfights13 package and the fights and planes tables to answer the following questions:

#2(a) (10 points) What is the oldest plane (specified by the tailnum variable) that flew from New York City airports in 2013?

In [9]:
q2a_1 = flights[["tailnum","dep_time"]].dropna(subset=["dep_time"])
q2a_2 = pd.merge(q2a_1,planes,on="tailnum",how="inner")
q2a_2.sort_values(by="year",ascending=True).head(5)

tailnum  dep_time    year                     type manufacturer  \
264146  N381AA    1505.0  1956.0  Fixed wing multi engine      DOUGLAS   
264129  N381AA    1525.0  1956.0  Fixed wing multi engine      DOUGLAS   
264130  N381AA    1737.0  1956.0  Fixed wing multi engine      DOUGLAS   
264131  N381AA     817.0  1956.0  Fixed wing multi engine      DOUGLAS   
264132  N381AA    1528.0  1956.0  Fixed wing multi engine      DOUGLAS   

         model  engines  seats  speed         engine  
264146  DC-7BF        4    102  232.0  Reciprocating  
264129  DC-7BF        4    102  232.0  Reciprocating  
264130  DC-7BF        4    102  232.0  Reciprocating  
264131  DC-7BF        4    102  232.0  Reciprocating  
264132  DC-7BF        4    102  232.0  Reciprocating

#2(b) (10 points) How many airplanes that flew from New York City are included in the planes table?

In [ ]:
q2b = set(q2a_2.tailnum)
len(q2b)

3316

In [ ]:
planes["tailnum"].count()

3322

**# 3316 planes that flew from New York City are included in the planes table. **

#3. Generate the code to convert the following data frame to wide format.
  grp sex meanL sdL meanR sdR
## 1 A F 0.225 0.106 0.340 0.0849
## 2 A M 0.470 0.325 0.570 0.3250
## 3 B F 0.325 0.106 0.400 0.0707
## 4 B M 0.547 0.308 0.647 0.2740

    grp F.meanL F.sdL F.meanR F.sdR M.meanL M.sdL M.meanR M.sdR
## <chr> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl> <dbl>
## 1 A   0.225 0.106 0.34 0.0849 0.47 0.325 0.57 0.325
## 2 B   0.325 0.106 0.4 0.0707 0.547 0.308 0.647 0.274

dat <- data.frame(grp = c("A","A","B","B"),
sex = c("F","M","F","M"),
meanL = c(0.225,0.47,0.325,0.547),
sdL = c(0.106,.325,.106,.308),
meanR = c(.34,.57,.4,.647),
sdR = c(0.0849, 0.325, 0.0707, 0.274)
)

In [ ]:


dat={"grp":["A","A","B","B"],"sex":["F","M","F","M"],"meanL":[0.225,0.47,0.325,0.547],"sdL":[0.106,.325,.106,.308],"meanR":[.34,.57,.4,.647],
     "sdR":[0.0849, 0.325, 0.0707, 0.274]}
dat_narrow = pd.DataFrame(dat)
dat_narrow

grp sex  meanL    sdL  meanR     sdR
0   A   F  0.225  0.106  0.340  0.0849
1   A   M  0.470  0.325  0.570  0.3250
2   B   F  0.325  0.106  0.400  0.0707
3   B   M  0.547  0.308  0.647  0.2740

In [ ]:
dat_wide = pd.pivot(dat_narrow,index="grp",columns="sex",values=["meanL","sdL","meanR","sdR"])
dat_wide

meanL           sdL        meanR            sdR       
sex      F      M      F      M     F      M       F      M
grp                                                        
A    0.225  0.470  0.106  0.325  0.34  0.570  0.0849  0.325
B    0.325  0.547  0.106  0.308  0.40  0.647  0.0707  0.274

#4. Consider the pccc icd10 dataset.
#(a) Remove all the columns labeled with "g" and a number.
#(b) Convert this to a long data set with three columns: id, type (pc or dx), and code.


In [ ]:
pccc.head(5)

id      dx1      dx2      dx3      dx4      dx5      dx6      dx7      dx8  \
0   1  S9410XS   I67841   E70339      NaN  S14121A   M66229  S92065G    O0973   
1   2      NaN  S53422D  S92244B   M66342      NaN  S32442A  T1582XD  S72325C   
2   3      NaN  S91225S      NaN  W6119XD    C8397  M80819K  S72114R      NaN   
3   4  S7226XK    Y93G2    L0592   K08530      NaN  S62637D  T84612A      NaN   
4   5  S92246A    O4212    D2920  S42434S   F15980      NaN  S52572R  M8080XA   

       dx9  ...        g1        g2        g3        g4        g5        g6  \
0      NaN  ...  239196.0  672832.0  683784.0  757546.0       NaN  168052.0   
1  S52131B  ...  931331.0  404900.0  912213.0       NaN  964580.0  371556.0   
2  Y382X3D  ...  627455.0  638100.0  745829.0  843799.0  322975.0       NaN   
3      NaN  ...  809782.0  153243.0  413723.0  130995.0  211708.0  610135.0   
4  X731XXD  ...       NaN  636794.0       NaN  928572.0  930823.0  168586.0   

         g7        g8        g9       g10  
0  104625.0       NaN  850974.0       NaN  
1  778488.0  115827.0  440619.0  955264.0  
2       NaN  932106.0  289004.0  242699.0  
3       NaN  471383.0  191245.0  135116.0  
4  133292.0  699936.0  500743.0       NaN  

[5 rows x 31 columns]

In [ ]:
pccc=pccc.rename(columns={"dx":"dx1","pc":"pc1","g":"g1"})
pccc

id      dx1      dx2      dx3      dx4      dx5      dx6      dx7  \
0       1  S9410XS   I67841   E70339      NaN  S14121A   M66229  S92065G   
1       2      NaN  S53422D  S92244B   M66342      NaN  S32442A  T1582XD   
2       3      NaN  S91225S      NaN  W6119XD    C8397  M80819K  S72114R   
3       4  S7226XK    Y93G2    L0592   K08530      NaN  S62637D  T84612A   
4       5  S92246A    O4212    D2920  S42434S   F15980      NaN  S52572R   
..    ...      ...      ...      ...      ...      ...      ...      ...   
995   996      NaN  S12041S     K121  S0522XA  O4102X2      NaN  S72392A   
996   997  S43203S  O3130X2      NaN    M9959      NaN   M67372    Q7271   
997   998      NaN  S45802S  X718XXS  S6440XA   F19129   F14281  S0910XA   
998   999    C3481  S21252S      NaN   I69342  S15011D  S23420S  T8459XD   
999  1000     O070    Q6471      NaN      NaN  S61312S   Z13850  S82013F   

         dx8      dx9  ...        g1        g2        g3        g4        g5  \
0      O0973      NaN  ...  239196.0  672832.0  683784.0  757546.0       NaN   
1    S72325C  S52131B  ...  931331.0  404900.0  912213.0       NaN  964580.0   
2        NaN  Y382X3D  ...  627455.0  638100.0  745829.0  843799.0  322975.0   
3        NaN      NaN  ...  809782.0  153243.0  413723.0  130995.0  211708.0   
4    M8080XA  X731XXD  ...       NaN  636794.0       NaN  928572.0  930823.0   
..       ...      ...  ...       ...       ...       ...       ...       ...   
995   I87012  T484X3S  ...  249628.0  765806.0  316110.0  546058.0       NaN   
996     Z809      NaN  ...  592401.0  163020.0       NaN  211442.0       NaN   
997    M7661   M67823  ...  273991.0  900149.0  858752.0       NaN       NaN   
998  V752XXS  S06383S  ...  723506.0  406516.0       NaN  300396.0       NaN   
999  T59894A  S52033P  ...       NaN  978996.0  612678.0  167076.0  679505.0   

           g6        g7        g8        g9       g10  
0    168052.0  104625.0       NaN  850974.0       NaN  
1    371556.0  778488.0  115827.0  440619.0  955264.0  
2         NaN       NaN  932106.0  289004.0  242699.0  
3    610135.0       NaN  471383.0  191245.0  135116.0  
4    168586.0  133292.0  699936.0  500743.0       NaN  
..        ...       ...       ...       ...       ...  
995  411926.0  294180.0       NaN  526766.0       NaN  
996  137483.0  677113.0  119594.0  447687.0  964283.0  
997       NaN       NaN  656304.0  320501.0  747476.0  
998  752696.0  142784.0  349031.0  772007.0  787355.0  
999  357490.0  465670.0       NaN       NaN  497168.0  

[1000 rows x 31 columns]

In [ ]:
pccc1 = pccc.drop(pccc.columns[pccc.columns.str.contains('g')],axis=1)
pccc1

id      dx1      dx2      dx3      dx4      dx5      dx6      dx7  \
0       1  S9410XS   I67841   E70339      NaN  S14121A   M66229  S92065G   
1       2      NaN  S53422D  S92244B   M66342      NaN  S32442A  T1582XD   
2       3      NaN  S91225S      NaN  W6119XD    C8397  M80819K  S72114R   
3       4  S7226XK    Y93G2    L0592   K08530      NaN  S62637D  T84612A   
4       5  S92246A    O4212    D2920  S42434S   F15980      NaN  S52572R   
..    ...      ...      ...      ...      ...      ...      ...      ...   
995   996      NaN  S12041S     K121  S0522XA  O4102X2      NaN  S72392A   
996   997  S43203S  O3130X2      NaN    M9959      NaN   M67372    Q7271   
997   998      NaN  S45802S  X718XXS  S6440XA   F19129   F14281  S0910XA   
998   999    C3481  S21252S      NaN   I69342  S15011D  S23420S  T8459XD   
999  1000     O070    Q6471      NaN      NaN  S61312S   Z13850  S82013F   

         dx8      dx9  ...      pc1      pc2      pc3      pc4      pc5  \
0      O0973      NaN  ...  0PSH3CZ  0JPT3XZ  037906Z  0JHD3HZ  0KQ54ZZ   
1    S72325C  S52131B  ...  0DVM7DZ  0NRJ47Z  DWY48ZZ  0HRWX7Z  BP091ZZ   
2        NaN  Y382X3D  ...  0PBV4ZX  0XM20ZZ  0DWD4UZ  2W07XYZ  F0636ZZ   
3        NaN      NaN  ...  DDY37ZZ  07LL0CZ  0Y9930Z  037M3GZ  04100Z4   
4    M8080XA  X731XXD  ...  02UL4KZ  03VD0ZZ  02110K8  3E050HZ  3E0U0GB   
..       ...      ...  ...      ...      ...      ...      ...      ...   
995   I87012  T484X3S  ...  0HN0XZZ  047R3D6  005P4ZZ  0PW1XKZ  041H4K9   
996     Z809      NaN  ...      NaN  0DQ70ZZ  06UN0JZ  03Q73ZZ  07L74ZZ   
997    M7661   M67823  ...  05BB0ZZ  0NSB0ZZ  0D957ZX  0CWY7KZ  07UH47Z   
998  V752XXS  S06383S  ...      NaN  0R954ZZ      NaN  0RU537Z  0PU64JZ   
999  T59894A  S52033P  ...  06923ZZ  0X340ZZ  09503ZZ      NaN  05R50JZ   

         pc6      pc7      pc8      pc9     pc10  
0    0WPK3YZ  01B04ZX  0DWV07Z  09513ZZ  0V554ZZ  
1    0Y0H4JZ      NaN  0B9880Z      NaN      NaN  
2    0RUP37Z      NaN  0WCP8ZZ  0DUM4KZ  BN02ZZZ  
3        NaN  0SPG33Z  0TRC07Z  041M0KQ  DB10B8Z  
4        NaN  0SPQ30Z  0WWBXYZ      NaN  0SWN38Z  
..       ...      ...      ...      ...      ...  
995  4A0D73Z  0QUM07Z  07L43ZZ      NaN  BR49ZZZ  
996  0VC53ZZ  2W08X5Z  0D5V0ZZ  0B9H8ZZ  0T184KD  
997  00XL4ZG  0GB10ZX  0S963ZZ  F0BZ0KZ      NaN  
998  03U337Z  03BQ0ZX  02VR3CT      NaN  0N940ZZ  
999  0BH082Z  01913ZX  05SY4ZZ  F07C6GZ  0NS5XZZ  

[1000 rows x 21 columns]

# another way:

In [ ]:
pccc1 = pccc.drop(pccc.loc[:, pccc.columns[pccc.columns.str.startswith('g')]], axis=1)
pccc1

id      dx1      dx2      dx3      dx4      dx5      dx6      dx7  \
0       1  S9410XS   I67841   E70339      NaN  S14121A   M66229  S92065G   
1       2      NaN  S53422D  S92244B   M66342      NaN  S32442A  T1582XD   
2       3      NaN  S91225S      NaN  W6119XD    C8397  M80819K  S72114R   
3       4  S7226XK    Y93G2    L0592   K08530      NaN  S62637D  T84612A   
4       5  S92246A    O4212    D2920  S42434S   F15980      NaN  S52572R   
..    ...      ...      ...      ...      ...      ...      ...      ...   
995   996      NaN  S12041S     K121  S0522XA  O4102X2      NaN  S72392A   
996   997  S43203S  O3130X2      NaN    M9959      NaN   M67372    Q7271   
997   998      NaN  S45802S  X718XXS  S6440XA   F19129   F14281  S0910XA   
998   999    C3481  S21252S      NaN   I69342  S15011D  S23420S  T8459XD   
999  1000     O070    Q6471      NaN      NaN  S61312S   Z13850  S82013F   

         dx8      dx9  ...      pc1      pc2      pc3      pc4      pc5  \
0      O0973      NaN  ...  0PSH3CZ  0JPT3XZ  037906Z  0JHD3HZ  0KQ54ZZ   
1    S72325C  S52131B  ...  0DVM7DZ  0NRJ47Z  DWY48ZZ  0HRWX7Z  BP091ZZ   
2        NaN  Y382X3D  ...  0PBV4ZX  0XM20ZZ  0DWD4UZ  2W07XYZ  F0636ZZ   
3        NaN      NaN  ...  DDY37ZZ  07LL0CZ  0Y9930Z  037M3GZ  04100Z4   
4    M8080XA  X731XXD  ...  02UL4KZ  03VD0ZZ  02110K8  3E050HZ  3E0U0GB   
..       ...      ...  ...      ...      ...      ...      ...      ...   
995   I87012  T484X3S  ...  0HN0XZZ  047R3D6  005P4ZZ  0PW1XKZ  041H4K9   
996     Z809      NaN  ...      NaN  0DQ70ZZ  06UN0JZ  03Q73ZZ  07L74ZZ   
997    M7661   M67823  ...  05BB0ZZ  0NSB0ZZ  0D957ZX  0CWY7KZ  07UH47Z   
998  V752XXS  S06383S  ...      NaN  0R954ZZ      NaN  0RU537Z  0PU64JZ   
999  T59894A  S52033P  ...  06923ZZ  0X340ZZ  09503ZZ      NaN  05R50JZ   

         pc6      pc7      pc8      pc9     pc10  
0    0WPK3YZ  01B04ZX  0DWV07Z  09513ZZ  0V554ZZ  
1    0Y0H4JZ      NaN  0B9880Z      NaN      NaN  
2    0RUP37Z      NaN  0WCP8ZZ  0DUM4KZ  BN02ZZZ  
3        NaN  0SPG33Z  0TRC07Z  041M0KQ  DB10B8Z  
4        NaN  0SPQ30Z  0WWBXYZ      NaN  0SWN38Z  
..       ...      ...      ...      ...      ...  
995  4A0D73Z  0QUM07Z  07L43ZZ      NaN  BR49ZZZ  
996  0VC53ZZ  2W08X5Z  0D5V0ZZ  0B9H8ZZ  0T184KD  
997  00XL4ZG  0GB10ZX  0S963ZZ  F0BZ0KZ      NaN  
998  03U337Z  03BQ0ZX  02VR3CT      NaN  0N940ZZ  
999  0BH082Z  01913ZX  05SY4ZZ  F07C6GZ  0NS5XZZ  

[1000 rows x 21 columns]

#pd.melt(df, id_vars=, value_vars=, var_name=, value_name=, ignore_index=)

In [ ]:
pccc1_narrow = pd.melt(pccc1, id_vars="id", value_vars=["dx1","dx2","dx3","dx4","dx5","dx6","dx7","dx8","dx9","dx10","pc1","pc2","pc3","pc4","pc5","pc6","pc7","pc8","pc9","pc10"],var_name="type",value_name="code")
pccc1_narrow

id  type     code
0         1   dx1  S9410XS
1         2   dx1      NaN
2         3   dx1      NaN
3         4   dx1  S7226XK
4         5   dx1  S92246A
...     ...   ...      ...
19995   996  pc10  BR49ZZZ
19996   997  pc10  0T184KD
19997   998  pc10      NaN
19998   999  pc10  0N940ZZ
19999  1000  pc10  0NS5XZZ

[20000 rows x 3 columns]

#another method:

In [ ]:
pccc2=pccc1

In [ ]:
pccc2.columns=pccc2.columns.str.rstrip("1,2,3,4,5,6,7,8,9,10")

In [ ]:
pccc2_long = pd.melt(pccc2,id_vars="id",value_vars=["dx","pc"],var_name="type",value_name="code")
pccc2_long

id type     code
0         1   dx  S9410XS
1         2   dx      NaN
2         3   dx      NaN
3         4   dx  S7226XK
4         5   dx  S92246A
...     ...  ...      ...
19995   996   pc  BR49ZZZ
19996   997   pc  0T184KD
19997   998   pc      NaN
19998   999   pc  0N940ZZ
19999  1000   pc  0NS5XZZ

[20000 rows x 3 columns]